convert plugin to fedora messaging

k-e-ian opened the pull request.

k-e-ian committed:
convert plugin to fedora messaging

abompard commented:
Very impressive, it's almost something I could have written myself... ;-)
Could you explain where you got your inspiration?

k-e-ian commented:
Thank you, Abompard!
Of course. I may have spent enough time going through this https://fedora-messaging.readthedocs.io/en/stable/user-guide/publishing.html for a better understanding of the fedora messaging object and functions (methods that come with).
I have also spent enough time going through your [Hacking on Fedora Infra apps] playlist on youtube for the Inspiration

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
         """
         return None
 
     def permalink(self, mlist, msg):
```
Method definition identical to https://github.com/fedora-infra/mailman3-fedmsg-plugin/blob/fedora-messaging/mailman3_fedmsg_plugin/__init__.py#L95

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
+        fm_msg = Message(body=body, topic="mailman.receive")
+        safe_publish(fm_msg)
 
     def list_url(self, mlist):
```
Method definition identical to https://github.com/fedora-infra/mailman3-fedmsg-plugin/blob/fedora-messaging/mailman3_fedmsg_plugin/__init__.py#L89

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
         "user-agent",
     ]
 
-    def __init__(self):
-        if not getattr(getattr(fedmsg, '__local', None), '__context', None):
-            hostname = socket.gethostname().split('.')[0]
-            fedmsg.init(name="mailman.%s" % hostname)
-        self.config = fedmsg.config.load_config()
-
-
```
Any reason why the class constructor was gotten rid of?

k-e-ian replied:
In the original implementation of the Archiver class, the constructor was used to initialize the fedmsg library, load configuration, and perform other setup tasks.
I got rid of it because we wont be needing it, fedora messaging will start the connection, in the first Message.

k-e-ian replied:
Do you suggest that it should remain?

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
-        self.config = fedmsg.config.load_config()
-
-
     def archive_message(self, mlist, msg):
```
Method definition identical to https://github.com/fedora-infra/mailman3-fedmsg-plugin/blob/fedora-messaging/mailman3_fedmsg_plugin/__init__.py#L101

k-e-ian replied:
the changes I made in the archive_message method are as follows:
Switched from fedmsg.publish to fedora_messaging.api.publish 
msg_metadata and lst_metadata remain the same.
Message object is from fedora_messaging.message.
I may have left out changing the comments under the archive_message referring to publishing to fedora messaging publisher rather than fedmsgbus.
Also the variable to fd_msg from msg.

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
-
-        fedmsg.publish(topic='receive', modname='mailman',
-                       msg=dict(msg=msg_metadata, mlist=lst_metadata))
+        def _format(value):
```
Method definition identical to https://github.com/fedora-infra/mailman3-fedmsg-plugin/blob/fedora-messaging/mailman3_fedmsg_plugin/__init__.py#L40

gridhead commented on mailman3_fedmsg_plugin.py:
```diff
@@ -67,47 +78,44 @@ class Archiver(object):
+        def _format(value):
+            return str(value) if value else value
+
+        msg_metadata = {key: _format(getattr(msg, key, None)) for key in self.keys}
+        lst_metadata = {
+            "list_name": mlist.list_name,
+            "mail_host": mlist.mail_host,
+            "fqdn_listname": mlist.fqdn_listname,
+            "list_id": mlist.list_id,
+            "display_name": mlist.display_name,
+        }
+        body = dict(msg=msg_metadata, mlist=lst_metadata)
+        fm_msg = Message(body=body, topic="mailman.receive")
+        safe_publish(fm_msg)
```
What method or section is this block under?

k-e-ian replied:
this section is encapsulated within the archive_message method and is responsible for the core functionality of archiving messages using Fedora Messaging.
body: It constructs the body of the message to be published to the Fedora Messaging bus. It includes both msg_metadata and lst_metadata dictionaries.

fm_msg: It creates a Message object using the body constructed in the previous step. This message object will be published to the Fedora Messaging bus with the topic "mailman.receive".

gridhead reviewed (changes requested):
Did you include tests for the additions that you have made?

abompard commented:
> Thank you, Abompard! Of course. I may have spent enough time going through this https://fedora-messaging.readthedocs.io/en/stable/user-guide/publishing.html for a better understanding of the fedora messaging object and functions (methods that come with). I have also spent enough time going through your [Hacking on Fedora Infra apps] playlist on youtube for the Inspiration

Yes, the code is *extremely* similar to what I wrote for the livestream. I still have it on my laptop. It's good that you found the livestreams and that you went through them, but I would have appreciated a mention that the pull request is 90% from there. Reusing other people's code is fine if the license allows it but attribution is always a good thing. Especially if your reviewer is the one who wrote the code in the first place (people tend to recognize their own code).

k-e-ian commented:
> > Thank you, Abompard! Of course. I may have spent enough time going through this https://fedora-messaging.readthedocs.io/en/stable/user-guide/publishing.html for a better understanding of the fedora messaging object and functions (methods that come with). I have also spent enough time going through your [Hacking on Fedora Infra apps] playlist on youtube for the Inspiration
> 
> Yes, the code is _extremely_ similar to what I wrote for the livestream. I still have it on my laptop. It's good that you found the livestreams and that you went through them, but I would have appreciated a mention that the pull request is 90% from there. Reusing other people's code is fine if the license allows it but attribution is always a good thing. Especially if your reviewer is the one who wrote the code in the first place (people tend to recognize their own code).

Thank you for the feedback, I did mention spending enough time going through YOUR [Hacking on Fedora Infra apps] playlist on youtube for the Inspiration. And was hoping to use that as a starting point, especially because that particular playlist made me more familiar with the codebase, than any other resource out there. 

I was hoping to build up from there, like finishing with the test units to confirm the issue has been completely done.

I realize now that I could have communicated this more clearly, and I apologize if it appeared that I was not properly acknowledging your contributions, especially considering that we will be working together.

k-e-ian commented:
> Did you include tests for the additions that you have made?

No I haven't, I tested it locally using RabbitMQ message broker, a consumer.py file that defines a callback and uses the fedora messaging api method to consume the message. And that was successful.

abompard commented:
Alright @k-e-ian , could you please add unit tests?

k-e-ian committed:
adding tests dir

k-e-ian commented:
> Alright @k-e-ian , could you please add unit tests?

Im still working on it, the test is failing. 
Im positive, in about <3 hours, ill have addressed your suggestions. Including @gridhead's.

k-e-ian commented:
> > Alright @k-e-ian , could you please add unit tests?
> 
> Im still working on it, the test is failing. Im positive, in about <3 hours, ill have addressed your suggestions. Including @gridhead's.

definitely underestimated the timing.

abompard commented:
Superceeded by b8763f842686aa348ff156370290162820dbad4b
